### PR TITLE
portfile-gen: new port

### PR DESCRIPTION
**Pull request by cooljeanius**
Closes: https://trac.macports.org/ticket/37797

#### Description

This is basically just the same Portfile that I submitted in Trac 37797, with the `# $Id$` line removed, and `revision` reset to `0`.

###### Type(s)
submission

###### Tested on
macOS 11.7.10 20G1427 x86_64
Xcode 13.2.1 13C100

###### Verification
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by reneeotten** [comment]
@cooljeanius is there really a need to add a port after 11 years of opening a Trac ticket. There are much better tools available nowadays (i.e., `upt`) for creating Portfile (semi-automatically) or the `cargo2port`, `cpan2port`, `go2port`, `pypi2port`. 

As one example, that script you're installing puts in `python.versions     25 26 27` for Python ports.... 

I am not in favor of merging this PR.

**Comment by reneeotten** [comment]
> > @cooljeanius is there really a need to add a port after 11 years of opening a Trac ticket.
> 
> Yes. If a ticket is still open, then it still needs to be fixed, regardless of the age.

It can also just be closed as "wont-fix" as I doubt anyone is waiting for this solution

**Comment by cooljeanius** [comment]
> > > @cooljeanius is there really a need to add a port after 11 years of opening a Trac ticket.
> > 
> > 
> > Yes. If a ticket is still open, then it still needs to be fixed, regardless of the age.
> 
> It can also just be closed as "wont-fix" as I doubt anyone is waiting for this solution

I am. I still use `portfile-gen`; all the other generators you mentioned are just for importing from other package managers. None of them handle the case of where you might want to start completely from scratch.

**Comment by pmetzger** [comment]
_If_ this is useful, then it has to be pulled from somewhere other than svn.macports.org because that thing is very dead. _If_ it is still useful, we do have a contrib scripts section in the main macports repos.

**Comment by cooljeanius** [comment]
> _If_ this is useful, then it has to be pulled from somewhere other than svn.macports.org because that thing is very dead. _If_ it is still useful, we do have a contrib scripts section in the main macports repos.

Yes, and this script is in that contrib scripts repo: https://github.com/macports/macports-contrib/blob/master/portfile-gen/portfile-gen
@jmroot updated it as recently as 5 months ago. I'll update this Portfile to pull from there instead.

File: sysutils/portfile-gen/Portfile
Comment (herbygillot) [inline sysutils/portfile-gen/Portfile:17]: This looks like a candidate for the `github` portgroup... ?

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:17]: OK so I'm trying to convert it to use that, but unfortunately the most recent release in the macports-contrib repo has a space in its name, which is giving me trouble: https://github.com/macports/macports-contrib/releases/tag/pypi2port-20230621

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:17]: wait actually nvm, I figured it out (pasting that url here helped!)

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile]: ok, shifted to use the `github` portgroup in [d88d81e](https://github.com/macports/macports-ports/pull/20951/commits/d88d81e5de6ff7dafcc2b60d466ed9cb9d06299c).

File: sysutils/portfile-gen/Portfile
Comment (jmroot) [inline sysutils/portfile-gen/Portfile:6]: This tag does technically point to the same commit as the currently most recent change to portfile-gen, but it might be less confusing to just use the commit (as future updates are unlikely to coincide with pypi2port release tags).

Also should use `github.tarball_from archive`.

File: sysutils/portfile-gen/Portfile
Comment (jmroot) [inline sysutils/portfile-gen/Portfile:13]: Should be `any`.

File: sysutils/portfile-gen/Portfile
Comment (jmroot) [inline sysutils/portfile-gen/Portfile:19]: Should use the `{*}` operator.

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:6]: oh right I forgot that commit hashes were valid to use as version numbers; so it'd be `5447906b1c3be5b6189e61c6256dc75c291c59e5` in this case?

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:19]: Uh...? I thought that was just for replacing the old `glob` syntax?

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile]: ok, changed to the hash in [cba79b7](https://github.com/macports/macports-ports/pull/20951/commits/cba79b7f4ec177d2b63630c0a4e76747d41106ca).

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile]: Fixed in [cba79b7](https://github.com/macports/macports-ports/pull/20951/commits/cba79b7f4ec177d2b63630c0a4e76747d41106ca).

File: sysutils/portfile-gen/Portfile
Comment (jmroot) [reply sysutils/portfile-gen/Portfile:20]: It doesn't replace `glob`, what it does it completely different. It does replace `eval` (as formerly used for compatibility Tcl < 8.5) in some cases like this where you want to append each element of a list to another list rather than appending the entire list as a single element.

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile]: > It doesn't replace `glob`, what it does it completely different. It does replace `eval` (as formerly used for compatibility with Tcl < 8.5) in some cases like this where you want to append each element of a list to another list rather than appending the entire list as a single element.

So, you mean like in [8491583](https://github.com/macports/macports-ports/pull/20951/commits/84915836d71363d864d0f0cf6a5d0c243e2d2d89)?

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile]: > > It doesn't replace `glob`, what it does it completely different. It does replace `eval` (as formerly used for compatibility with Tcl < 8.5) in some cases like this where you want to append each element of a list to another list rather than appending the entire list as a single element.
> 
> So, you mean like in [8491583](https://github.com/macports/macports-ports/pull/20951/commits/84915836d71363d864d0f0cf6a5d0c243e2d2d89)?

...assuming so; marking as resolved.

**Comment by pmetzger** [comment]
FWIW, I'm waiting for someone like @jmroot to make a decision on what to do here.

**Comment by subpop** [comment]
As a fairly new port maintainer, I would find a tool like this useful. And as a fairly new port maintainer, I am more likely to discover this utility by searching the ports database than finding the macports-contrib git repository.

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [inline sysutils/portfile-gen/Portfile:21]: same as in the PR I just merged, this could be simplified in the same way as there

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [inline sysutils/portfile-gen/Portfile:25]: this shouldn't be needed - why do you use this?

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [inline sysutils/portfile-gen/Portfile:23]: why would you switch off the livecheck?

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [inline sysutils/portfile-gen/Portfile:32]: if this is implied then there is no need to spell this out, or...?

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [inline sysutils/portfile-gen/Portfile:36]: same as in the other PR - remove useless message

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [inline sysutils/portfile-gen/Portfile:39]: same comment as in the other PR I just merged - please update

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:32]: I think it's better to be explicit than implicit.

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:36]: It's not useless, it's helpful and informative. All ports that override build steps should explicitly tell me when they're doing so, IMO.

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:39]: done in 7907811

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:21]: done in 7907811

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:25]: Probably a leftover from back when it still was needed; removed in 7907811

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:23]: Probably a leftover from when I couldn't get it to work previously; removed in 7907811

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:36]: Added an extra comment in 7907811 telling people not to remove it.

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [reply sysutils/portfile-gen/Portfile:32]: Please remove as requested

File: sysutils/portfile-gen/Portfile
Comment (reneeotten) [reply sysutils/portfile-gen/Portfile:37]: Please remove as requested

**Review by reneeotten — CHANGES_REQUESTED** [review]
Final attempt to get this merged. Please remove the redundant/unnecessary pieces in the portfile as requested here (and in other PRs if memory serves me). That's the only way I will merge it, otherwise I'll close the PR.

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:37]: It's necessary to prevent that block from being empty, which is something that's warned about in other programming languages, see for example the gcc/clang warning `-Wempty-body` for the C family of languages, or luacheck's W542 for lua. It's also necessary to avoid violating the principle of least surprise; a user may wonder why a step is getting overridden, and this makes it clear what's going on by explaining it.

**Comment by cooljeanius** [comment]
> Final attempt to get this merged. Please remove the redundant/unnecessary pieces in the portfile as requested here (and in other PTs if memory serves me). That's the only way I will merge it, otherwise I'll close the PR.

They're not redundant/unnecessary; they're essential for helping me understand why the port works the way it does. If you won't merge it, then let's get a different maintainer who isn't needlessly picky about Portfile authors' unique coding styles to merge it instead.

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile]: OK fine, I gave in on this one in [c206e18](https://github.com/macports/macports-ports/pull/20951/commits/c206e18251b8972f8370ff8cbf30d3198b5a5be0).

File: sysutils/portfile-gen/Portfile
Comment (pmetzger) [reply sysutils/portfile-gen/Portfile:34]: This is not done in other Portfiles, and there is no need for it. The fact that some programming languages warn about empty blocks isn't really germane.

**Comment by reneeotten** [comment]
well, this has been in the queue for 8 months without any attention so there appears very little appetite to get this merged at all. 

I've made my point, as have others here, so without changes by the end of the weekend I'll close the PR. You can feel free to have this in your local ports tree

File: sysutils/portfile-gen/Portfile
Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:34]: Well it *ought* to be done in other Portfiles. There is a desire for it.

Comment (pmetzger) [reply sysutils/portfile-gen/Portfile:34]: Maintaining MacPorts to a high degree of quality requires a lot of work on the part of people like @reneeotten who merge huge numbers of pull requests. It is necessary to have uniformity and reasonably standard practice in order to make review fast and efficient. Something that only adds a couple of minutes to a review adds hours to his monthly burden. So it does matter.

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:34]: > Maintaining MacPorts to a high degree of quality requires a lot of work on the part of people like @reneeotten who merge huge numbers of pull requests. It is necessary to have uniformity and reasonably standard practice in order to make review fast and efficient. Something that only adds a couple of minutes to a review adds hours to his monthly burden. So it does matter.

Just say yes, it's not that hard.

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:34]: > It's easy to make work for other people, it's a lot harder to do that work.

Clicking the "approve" button takes no more than a second, less time than I spent writing the line in question. Also I can help review PRs if that's what you need.

Comment (pmetzger) [reply sysutils/portfile-gen/Portfile:34]: You can't just click approve, you have to actually go through and make sure that the thing _can_ be approved because it's correct. The simpler, more stereotypical, and easier to understand a request is, the less time it takes to figure out if you can approve it or not. This is why we need Portfiles to be simple, comprehensible, and to follow stereotyped patterns.

Comment (cooljeanius) [reply sysutils/portfile-gen/Portfile:34]: It *is* correct, though. I don't see what's difficult to understand about a single `ui_debug` statement; that's the stereotypical pattern for providing extra information that only some users might want to see.

**Pull request closed without merging**
